lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java

```diff
@@ -231,6 +261,97 @@ public APIGatewayProxyResponseEvent handleRequest(
         }
     }
 
+    private static Optional<InheritedIdentityJwtClaim> getInheritedIdentityClaim(
+            JWTClaimsSet claimsSet) throws ParseException, RecoverableJarValidationException {
+        try {
+            return Optional.ofNullable(
+                            OBJECT_MAPPER.convertValue(
+                                    claimsSet.getJSONObjectClaim(CLAIMS_CLAIM), JarClaims.class))
+                    .map(JarClaims::userInfo)
+                    .map(JarUserInfo::inheritedIdentityClaim);
+        } catch (IllegalArgumentException | ParseException e) {
+            throw new RecoverableJarValidationException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT.setDescription(
+                            "Claims cannot be parsed to JarClaims"),
+                    claimsSet,
+                    e);
+        }
+    }
+
+    @Tracing
+    private void validateAndStoreHMRCInheritedIdentity(
+            String userId,
+            InheritedIdentityJwtClaim inheritedIdentityJwtClaim,
+            JWTClaimsSet claimsSet,
+            IpvSessionItem ipvSessionItem)
+            throws RecoverableJarValidationException, ParseException {
+        try {
+            var signedInheritedIdentityJWT =
+                    validateHmrcInheritedIdentity(userId, inheritedIdentityJwtClaim);
+            storeHmrcInheritedIdentity(
+                    userId, claimsSet, ipvSessionItem, signedInheritedIdentityJWT);
+        } catch (VerifiableCredentialException | ParseException e) {
+            throw new RecoverableJarValidationException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT.setDescription(
+                            "Inherited identity JWT failed to validate"),
+                    claimsSet,
+                    e);
+        } catch (JarValidationException e) {
+            throw new RecoverableJarValidationException(e.getErrorObject(), claimsSet, e);
+        }
+    }
+
+    private SignedJWT validateHmrcInheritedIdentity(
+            String userId, InheritedIdentityJwtClaim inheritedIdentityJwtClaim)
+            throws JarValidationException, ParseException, VerifiableCredentialException {
+        // Validate JAR claims structure is valid
+        var inheritedIdentityJwtList =
+                Optional.of(inheritedIdentityJwtClaim.value())
```

Comment:
You need to use `ofNullable` if you think it may be nullable (`of` will throw an NPE)